CodeInput/JsonInput should format automatically in readOnly mode.

RyanMG opened the pull request:
Linked issue: https://github.com/xh/hoist-react/issues/1923

### What

Adds an `autoFormat` prop to `CodeInput` / `JsonInput`. When set, the input formats its content for display via the configured `formatter`:

- **Readonly inputs** - defaults to `true` (pass `false` to opt out). Callers can bind raw values and let the input render them.
- **Editable inputs** - opt in with `autoFormat: true` to format content on blur (never mid-edit, so user edits and cursor position are preserved while typing). Users get clean, consistently-formatted JSON without reaching for the format button.

The `formatter` is invoked within a try/catch, so a formatter that throws (e.g. on invalid JSON) leaves the original content unchanged.

### Why

Callers no longer need to pre-format JSON before handing it to a `jsonInput`. For readonly displays they can pass the raw value (object or string) and let the input render it - since `fmtJson` delegates to `JSON.stringify`, an object's `toJSON()` is honored automatically. This makes display code more declarative (value = data, `formatter` = how to render) and removes a pile of inline `fmtJson(...)` / `JSON.stringify(..., null, 2)` calls across the Admin Console.

The editable case is an equally meaningful win: writable JSON editors stay tidy on their own. Enabled on the two library writable editors - the grid filter dialog (`GridFilterDialog`) and the REST form `json` field (`RestFormField`) - dropping the manual pretty-printing in the former.

Admin JSON panels are consolidated behind a small shared `adminJsonDisplay` component.

Companion Toolbox PR: https://github.com/xh/toolbox/pull/869

Hoist P/R Checklist
-------------------

- [X] Caught up with `develop` branch as of last change.
- [X] Added CHANGELOG entry, or determined not required.
- [X] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [X] Updated doc comments / prop-types, or determined not required.
- [X] Reviewed and tested on Mobile, or determined not required.
- [X] Created Toolbox branch / PR, or determined not required.

RyanMG committed:
feat(jsonInput): format the provided value to display properly

 - prevents need for each use of jsonInput to manually format the value

RyanMG committed:
refactor(jsonInput): remove auto formatting of JSON data for connection pool monitor

 - will be formatted in the JSONinput component itself

RyanMG commented on desktop/cmp/input/JsonInput.ts:
```diff
@@ -21,11 +21,18 @@ export const [JsonInput, jsonInput] = hoistCmp.withFactory<JsonInputProps>({
     render(props, ref) {
+        let {value, ...jsonInputProps} = props;
+
+        if (typeof value !== 'string') {
```
Originally was checking if value was an object, but `null` seems to be a valid value to be passed in, and is an object in JS.

RyanMG committed:
post-pr review updates to jsonInput formatting update

 - use lodash for data type checks (rather than typeof)
 - rename spread props to reference that they are going INTO codeInput, rather than OUT OF jsonInput

RyanMG committed:
doc: update CHANGELOG for the jsonInput format update

haynesjm42 commented on desktop/cmp/input/JsonInput.ts:
```diff
@@ -21,11 +22,20 @@ export const [JsonInput, jsonInput] = hoistCmp.withFactory<JsonInputProps>({
+        let {value, ...codeInputProps} = props;
+
+        if (isPlainObject(value)) {
+            value = JSON.stringify(value, null, 2);
```
Any processing of the value should be done in the input model, in this case that would be `CodeInputModel` since this input control is just a wrapper on the `codeInput` component, might make sense to introduce a `JsonInputModel` class here for special json handling. `HoistInputModel` is the base class for all input component models and has methods for converting between the external and internal values, which is where code like this needs to live. It is important to use those methods as our input components support both the controlled mode (providing `value` and `onChange` in props) as well as binding to a property of a provided model.

`DateInput` is probably a good example to follow here, where it has a `valueType` prop to tell it whether it should be dealing with a `LocalDate` or a `Date`.

I'm personally in favor of this change overall, others might feel differently as it does open the door for potentially confusing behavior as the object will be stringified and then parsed when written back, though any usages currently would effectively be doing the same thing, just that they'd just be needing to do the stringifying and parsing themselves.

@amcclain / @lbwexler any thoughts on supporting both strings and objects in JsonInput?

amcclain replied:
It's a good question - I had not considered it.

My concern was that it could muddy the waters a bit in terms of what value types this component supports. Was thinking we could use a type to represent valid JSON values - typefest has `JsonValue` which I believe would be it.

However for that to be a valid concern at all, we'd need to actually spec the type of values that inputs support, and as David points out above, `HoistInputProps.value` is stuck on `any`.  Can we add a generic to support setting the value type on that props interface?  

That would be a very different change from pretty-printing - I think that still need to be addressed, see comment on the linked issue, but adding proper value typing _seems_ like a good idea.  My only concern is that if we did so we would create work for devs who were unwittingly benefitting from looser typing with `any`, but that's kinda the idea...

amcclain commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@
+* Updated the `jsonInput` component to auto-format its provided value for display.
+It can accept JSON, stringified JSON or plain objects and output proper pretty-printed JSON in the
+UI - without the need for manual formatting in the call.
+
```
We had a release underneath this PR, sorry :) - will need to move this up

RyanMG committed:
codeInput autoformatting update - prop to auto format the display when showing readonly values

 - specific use case here is pretty printing JSON data without manual fromatting of provided values
 - codeInput does all the heavy lifting, so change is there (possible use cases for other formats)
 - update to admin activity details use of jsonInput to test change

RyanMG committed:
clean up activity detail admin JSON display

 - no longer attempts to format the JSON for display manually
 - rename an observed prop to represent that it is not auto-formatted

RyanMG committed:
Merge branch 'develop' into remove-json-preformat

RyanMG committed:
doc: CHANGELOG update to account for version bump

RyanMG committed:
revert original batch of jsonInput changes

 - includes an update to a component which was passing a raw object into jsonInput
 - objects will still need to be stringified to use in jsonInput / codeInput
 - also includes doc update to the codeInput interface

RyanMG commented on admin/tabs/activity/tracking/detail/ActivityDetailModel.ts:
```diff
@@ -9,15 +9,14 @@ import * as Col from '@xh/hoist/admin/columns';
     @managed gridModel: GridModel;
     @managed formModel: FormModel;
-    @observable formattedData;
+    @observable displayData;
```
This model is no longer doing the formatting. Update the name to reflect. See https://github.com/xh/hoist-react/pull/3937/files#diff-2a82d78cc59aadd90860fcd0383be76d3b7491d7b19736a18857b6b93d30cd66L117-L125

RyanMG committed:
have the default formatter for jsonInput be the safe version of the JSON formatter

RyanMG commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -326,22 +334,33 @@ class CodeInputModel extends HoistInputModel {
         }
     }
 
+    private formatValue() {
```
@amcclain had some naming / formatting updates he wanted done here.

1. onAutoFormat -> formatValue. We are choosing to use autoFormat as the new prop (on interface) to format readonly data when loading into the UI. Anselm felt onAutoFormat was not the best name of the reaction happening here (to a user requesting the use of the formatter to update their UI)
2.  tryPrettyPrint was use only once, in this method, to run the formatter. Inlining that call to the method itself.

RyanMG commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -326,22 +334,33 @@ class CodeInputModel extends HoistInputModel {
 
-    onAutoFormat = () => {
-        if (!isFunction(this.componentProps.formatter)) return;
+    override toInternal(val: any) {
```
Override fo the toInternal used by the `HoistInputModel`. Now will run the formatter if requested and properly configured.

RyanMG committed:
method name update from formatValue -> formatAndSetEditorValue to better describe the functionality occuring

RyanMG commented on admin/tabs/activity/tracking/detail/ActivityDetailView.ts:
```diff
@@ -134,10 +134,11 @@ const detailRecForm = hoistCmp.factory<ActivityDetailModel>(({model}) => {
         height: '100%',
         showCopyButton: true,
-        value: model.formattedData
+        value: model.displayData
```
See https://github.com/xh/hoist-react/pull/3937/files#r2001715715

RyanMG commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -326,22 +334,33 @@ class CodeInputModel extends HoistInputModel {
         }
     }
 
+    private formatAndSetEditorValue() {
```
@amcclain had some naming / formatting updates he wanted done here.

onAutoFormat -> formatValue was the first pass at the name. We are choosing to use autoFormat as the new prop (on interface) to format readonly data when loading into the UI. Anselm felt onAutoFormat was not the best name of the reaction happening here (to a user requesting the use of the formatter to update their UI)
tryPrettyPrint was use only once, in this method, to run the formatter. Inlining that call to the method itself.
However, we subsequently added safeFormatValue, which is much leaner - it just takes a val, tries to format it, and returns it one way or the other. This existing method was already safe (try/catch block), so to a future developer it may be configuring why the similar names. So I opted to update this method to denote that it would be sending the value to the editor (code mirror) instead of returning said value.

RyanMG commented on desktop/cmp/input/JsonInput.ts:
```diff
@@ -22,8 +22,8 @@ export const [JsonInput, jsonInput] = hoistCmp.withFactory<JsonInputProps>({
-            linter: linter,
-            formatter: fmtJson,
+            linter,
+            formatter: safeFmtJson,
```
Unsure if we want this? We create a "safe" version of `fmtJson` (https://github.com/xh/hoist-react/pull/3937/files#diff-1bd2a34b6889e1254a813e22f3fecee2169724007549f9c6d2396658c54ebc7bR76). I made this update to use that for jsonInput, but that may create issues I am not thinking of?

RyanMG commented:
Had a meeting this morning with @amcclain, and in that we reviewed this issue broadly. Updates done in that meeting applied to this PR. CC:// @haynesjm42

RyanMG commented on desktop/cmp/input/CodeInput.ts:
```diff
@@ -43,6 +43,14 @@ export interface CodeInputProps extends HoistProps, HoistInputProps, LayoutProps
     /** True to focus the control on render. */
```
This is a `desktop` component. I did not see a mobile version, but let me know if I just overlooked something.

RyanMG committed:
Merge branch 'develop' into remove-json-preformat

RyanMG committed:
Remove jest testing code which got stuck and commited on accident

RyanMG committed:
Merge branch 'develop' into remove-json-preformat

RyanMG committed:
Return some of the changes done by the linted automatically

lbwexler committed:
Merge remote-tracking branch 'origin/develop' into remove-json-preformat

# Conflicts:
#	desktop/cmp/input/CodeInput.ts
#	desktop/cmp/input/JsonInput.ts

lbwexler committed:
Add CHANGELOG entry for jsonInput autoFormat / safe formatting

lbwexler committed:
Merge remote-tracking branch 'origin/develop' into remove-json-preformat

lbwexler committed:
Default codeInput autoFormat on for readonly inputs; warn if set on editable

- autoFormat now defaults true when readonly (opt out with false)
- logWarn when autoFormat is set on a non-readonly input (no-op)
- Clarify autoFormat doc comment

lbwexler committed:
Drop redundant manual JSON pretty-printing for readonly jsonInputs

Readonly inputs now auto-format for display, so pre-formatting the value is
no longer needed. Left timestampReplacer-based formatting in place (DetailPanel,
DetailsPanel, ActivityDetail) - that is a real transform, not just pretty-print.

lbwexler committed:
Use warnIf for autoFormat-on-editable warning

lbwexler committed:
Make timestamp-formatted readonly jsonInputs declarative

Pass the raw stats object as value and move the timestampReplacer formatting
into a formatter prop, so autoFormat drives display formatting (DetailPanel,
service DetailsPanel).

lbwexler committed:
Drop redundant safeFmtJson; rely on CodeInput's formatter try/catch

The component already invokes the formatter within a try/catch (tryPrettyPrint),
so JsonInput can use fmtJson directly and any throwing formatter is handled in
the component - no separate safe wrapper needed.

lbwexler committed:
Drop redundant filter.toJSON() - JSON.stringify invokes it automatically

lbwexler committed:
Bind pool config jsonInput to poolConfiguration model observable

lbwexler committed:
Resolve pool config bind via context model lookup

lbwexler committed:
Extract shared adminJsonDisplay for readonly admin JSON panels

DRY up the repeated readonly jsonInput shell (full-size, no line numbers /
fullscreen button, timestamp-aware formatting) used by the cluster object
DetailPanel, service DetailsPanel, and ConnPoolMonitorPanel.

lbwexler committed:
Build flex: 1 into adminJsonDisplay default sizing

lbwexler committed:
Bind DetailPanel json display to selectedAdminStats

lbwexler committed:
Bind service DetailsPanel json display to stats

lbwexler committed:
checkpoint

lbwexler committed:
Note readonly JsonInput call-site simplification in CHANGELOG

lbwexler committed:
Merge remote-tracking branch 'origin/develop' into remove-json-preformat

lbwexler committed:
Allow autoFormat on editable CodeInput (formats on blur)

Previously autoFormat applied only to readonly inputs and warned when set
on editable ones. Because the format runs through toInternal - which only
feeds renderValue while unfocused - it never reformats mid-edit, so it is
safe to honor on editable inputs as a format-on-blur behavior.

- autoFormat getter now defaults to !!readonly, honoring an explicit true
  on editable inputs
- Dropped the warnIf no-op gate and the now-unused warnIf import
- Updated prop docs and CHANGELOG

lbwexler committed:
Enable autoFormat on editable JSON editors (grid filter, REST form)

Turn on the new format-on-blur behavior for the two writable library
jsonInputs, so users get clean, consistently-formatted JSON automatically:

- GridFilterDialog: enable autoFormat and drop the manual
  JSON.stringify(..., undefined, 2) pretty-printing in loadForm - fmtJson
  (the JsonInput default formatter, space: 2) now handles display formatting
- RestFormField: enable autoFormat on the json field editor (binding was
  already direct, nothing to unwind)
- Reframe the CHANGELOG entry to give the editable case equal billing

lbwexler committed:
tweaks from code review
